test/test_sparse.py

```diff
@@ -509,8 +565,121 @@ def _test_sparse_mask_fixed(self):
         expected = self.SparseTensor(i, exp_v, torch.Size([5, 4]))
         self.assertEqual(res, expected)
 
+        expected_dense = self.ValueTensor([
+            [1, 2, 300, 4],
+            [5, 6, 100, 8],
+            [9, 10, 11, 12],
+            [13, 200, 15, 16],
+            [17, 18, 19, 400],
+        ])
+        dense._sparse_copy_(x)
+        self.assertEqual(dense, expected_dense)
+
+    def prop_sparse_mask_const(self, x, s):
+        """sparse_mask does not mutate inputs."""
+        expected_x = x.clone()
+        expected_s = s.clone()
+        try:
+            x._sparse_mask(s)
+        except:
+            pass
+        self.assertEqual(x, expected_x)
+        self.assertEqual(s, expected_s)
+
+    def _sparse_mask_py(self, x, s):
+        """Python reference implementation of sparse_mask"""
+        if not s.is_coalesced():
+            raise ValueError("mask is not coalesced")
+        if x.size() != s.size():
+            raise ValueError("sparse_mask operands have incompatible sizes")
+        ss = SparseSize(size=x.size(), dimI=s._indices().size()[0], nnz=s._nnz())
+        r = ss.new_sparse(dense2sparse(type(x)), s._indices())
+        v = r._values()
+        for i, js in enumerate(s._indices().t()):
+            m = x
+            for j in js:
+                # NB: m[j] will happily accept a negative index; but sparse_mask
+                # really shouldnot!
+                if j < 0:
+                    raise IndexError("negative index is out of range")
+                m = m[j]
+            v[i] = m
+        return r
+
+    def prop_sparse_mask_py(self, x, s):
+        """sparse_mask agrees with reference implementation"""
+        # NB: It would be better if we could make more precise statements
+        # about what exceptions should be raised if there is an error,
+        # but the different APIs are actually quite inconsistent here.
+        expect_error = False
+        error = False
+        expected_r = None
+        try:
+            expected_r = self._sparse_mask_py(x.clone(), s.clone())
+        except (ValueError, RuntimeError, IndexError) as e:
+            expect_error = e
+        try:
+            r = x._sparse_mask(s)
+        except (ValueError, RuntimeError, IndexError) as e:
+            error = e
+        if expect_error and not error:
```

Comment:
So, we can indeed rewrite this code as:

```
        if expect_error:
            self.assertRaises((ValueError, RuntimeError, IndexError), x._sparse_mask, s)
        else:
            self.assertEqual(x._sparse_mask(s), expected_r)
```

But I think this is slightly worse than the custom code I have. The problem is that assertRaises doesn't conveniently allow specifying a custom message, and I have found, in practice, that it is very helpful to see WHY we were expecting an error (e.g., what error was it that the reference implementation threw), because sometimes the reference implementation was wrong, and the C code was right.

You may also wonder why we don't assert that a specific exception is thrown. The answer is that the CUDA and C implementations of operators throw different errors for the same error conditions; in particular, C throws an IndexError on an OOB, while in CUDA this is an assert failure which shows up as a RuntimeError.